cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
+        the corresponding execution coroutine `_manage_request()` is notified.
+
+        When the stream breaks, the stream is reopened, and all execution coroutines are notified.
+
+        There is at most a single instance of this coroutine running.
+        """
+        while True:
+            try:
+                # The default gRPC client timeout is used.
+                response_iterable = await self._grpc_client.quantum_run_stream(
+                    self._request_iterator()
+                )
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except BaseException as e:
+                # TODO(#5996) Close the request iterator to close the existing stream.
+                # Note: the message ID counter is not reset upon a new stream.
+                self._response_demux.publish_exception(e)  # Raise to all request tasks
+                if isinstance(e, asyncio.CancelledError):
+                    break
+
+    async def _manage_execution(
+        self, project_name: str, program: quantum.QuantumProgram, job: quantum.QuantumJob
+    ) -> Union[quantum.QuantumResult, quantum.QuantumJob]:
+        """The execution coroutine, an asyncio coroutine to manage the lifecycle of a job execution.
+
+        This coroutine sends QuantumRunStream requests to the request iterator and receives
+        responses from the ResponseDemux.
+
+        It initially sends a CreateQuantumProgramAndJobRequest, and retries if there is a retryable
+        error by sending another request. The exact request type depends on the error.
+
+        There is one execution coroutine per running job submission.
+        """
+
+        # Construct requests ahead of time to be reused for retries.
+        create_program_and_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+        create_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_job=quantum.CreateQuantumJobRequest(
+                parent=program.name, quantum_job=job
+            ),
+        )
+        get_result_request = quantum.QuantumRunStreamRequest(
+            parent=project_name, get_quantum_result=quantum.GetQuantumResultRequest(parent=job.name)
+        )
+
+        current_request = create_program_and_job_request
+        while True:
+            try:
+                current_request.message_id = self._generate_message_id()
+                response_future = self._response_demux.subscribe(current_request.message_id)
+                await self._request_queue.put(current_request)
+                response = await response_future
+
+            # Broken stream
```

Comment:
Have you verified that broken streams look like this?  e.g. by terminating a locally running server or messing up your connection to a prod instance (or better yet - with an e2e test)?

Reply:
I have not, thanks for the great catch. Will do an experiment in parallel with this PR.